isic/zip_download/api.py

```diff
@@ -8,128 +8,106 @@
-from django.urls.base import reverse
-from drf_yasg.utils import swagger_auto_schema
-from rest_framework.decorators import api_view, permission_classes
-from rest_framework.exceptions import PermissionDenied
-from rest_framework.pagination import CursorPagination
-from rest_framework.permissions import AllowAny
-from rest_framework.response import Response
-
-from isic.core.models.base import CopyrightLicense
-from isic.core.models.image import Image
-from isic.core.serializers import SearchQuerySerializer
+from ninja import Query, Router
+
+from isic.core.models import CopyrightLicense, Image
+from isic.core.pagination import CursorPagination
+from isic.core.serializers import SearchQueryIn
 from isic.core.services import _image_metadata_csv_headers, image_metadata_csv_rows
-from isic.zip_download.serializers import ZipFileEntrySerializer
 
 logger = logging.getLogger(__name__)
+zip_router = Router()
 
 
 # this is directly mirrored in isic-cli
 def get_attributions(attributions: Iterable[str]) -> list[str]:
-    counter = Counter(attribution for attribution in attributions)
+    counter = Counter(attributions)
     # sort by the number of images descending, then the name of the institution ascending
     attributions = sorted(counter.most_common(), key=lambda v: (-v[1], v[0]))
     # push anonymous attributions to the end
     attributions = sorted(attributions, key=lambda v: 1 if v[0] == "Anonymous" else 0)
     return [x[0] for x in attributions]
 
 
-def get_zip_download_token(token: str | None = None) -> dict:
-    if not token:
-        raise PermissionDenied
-
-    signer = TimestampSigner()
-
-    try:
-        # TODO: do we sign tokens longer than URLs?
-        return signer.unsign_object(token, max_age=timedelta(days=1))
-    except BadSignature:
-        raise PermissionDenied
-
+def get_zip_download_token(token: str) -> dict:
+    return TimestampSigner().unsign_object(token, max_age=timedelta(days=1))
 
-@swagger_auto_schema(methods=["POST"], auto_schema=None)
-@api_view(["POST"])
-@permission_classes([AllowAny])
-def create_zip_download_url(request):
-    serializer = SearchQuerySerializer(data=request.data, context={"user": request.user})
-    serializer.is_valid(raise_exception=True)
 
-    token = TimestampSigner().sign_object(serializer.to_token_representation())
+@zip_router.post("/url/", response=str, include_in_schema=False)
+def create_zip_download_url(request: HttpRequest, payload: SearchQueryIn):
+    token = TimestampSigner().sign_object(payload.to_token_representation(user=request.user))
+    return f"{settings.ZIP_DOWNLOAD_SERVICE_URL}/download?zsid={token}"
 
-    return Response(f"{settings.ZIP_DOWNLOAD_SERVICE_URL}/download?zsid={token}")
 
+@zip_router.get("/file-listing/", include_in_schema=False)
+def zip_file_listing(
+    request: HttpRequest,
+    token: str = Query(..., min_length=1),  # noqa B008
+    pagination: CursorPagination.Input = Query(...),  # noqa B008
+):
+    try:
```

Comment:
This might be top notch pedantry, but does it make sense to use a https://django-ninja.rest-framework.com/guides/authentication/#custom-function for the zip token?

Reply:
Yeah, I quite like that. I'll do it.